File: app/src/main/java/at/shockbytes/coins/dagger/AppModule.java
package at.shockbytes.coins.dagger;

import android.app.Application;

import javax.inject.Named;
import javax.inject.Singleton;

import at.shockbytes.coins.currency.CurrencyManager;
import at.shockbytes.coins.currency.DefaultCurrencyManager;
import at.shockbytes.coins.currency.DefaultPriceProxy;
import at.shockbytes.coins.currency.PriceProxy;
import at.shockbytes.coins.network.PriceManager;
import at.shockbytes.coins.network.conversion.CurrencyConversionApi;
import at.shockbytes.coins.storage.CoinsRealmMigration;
import at.shockbytes.coins.storage.RealmStorageManager;
import at.shockbytes.coins.storage.StorageManager;
import dagger.Module;
import dagger.Provides;
import io.realm.Realm;
import io.realm.RealmConfiguration;

/**
 * @author Martin Macheiner
 *         Date: 14.06.2017.
 */
@Module
public class AppModule {

    private Application app;

    public AppModule(Application application) {
        this.app = application;
    }

    @Provides
    @Singleton
    public CurrencyManager provideCurrencyManager(PriceProxy priceProxy,
                                                  StorageManager storageManager,
                                                  CurrencyConversionApi currencyConversionApi) {
        return new DefaultCurrencyManager(app.getApplicationContext(), priceProxy,
                storageManager, currencyConversionApi);
    }

    @Provides
    @Singleton
    public PriceProxy providePriceProxy(@Named("CoinbasePriceManager") PriceManager priceManager) {
        return new DefaultPriceProxy(priceManager);
    }

    @Provides
    @Singleton
    public Realm provideRealm() {
        RealmConfiguration config = new RealmConfiguration.Builder()
                .schemaVersion(CoinsRealmMigration.CURRENT_SCHEME)
                .migration(new CoinsRealmMigration())
                //.encryptionKey(key)
                .build();
        return Realm.getInstance(config);
    }

    @Provides
    @Singleton
    public StorageManager provideStorageManager(Realm realm) {
        return new RealmStorageManager(realm);
    }

}

File: app/src/main/java/at/shockbytes/coins/dagger/AppModule.kt
package at.shockbytes.coins.dagger

import android.app.Application
import android.content.SharedPreferences
import android.preference.PreferenceManager
import at.shockbytes.coins.currency.CurrencyManager
import at.shockbytes.coins.currency.DefaultCurrencyManager
import at.shockbytes.coins.currency.conversion.CurrencyConversionProvider
import at.shockbytes.coins.currency.price.DefaultPriceProxy
import at.shockbytes.coins.currency.price.PriceProvider
import at.shockbytes.coins.currency.price.PriceProxy
import at.shockbytes.coins.storage.StorageManager
import at.shockbytes.coins.storage.realm.CryptoWatcherRealmMigration
import at.shockbytes.coins.storage.realm.RealmStorageManager
import dagger.Module
import dagger.Provides
import io.realm.Realm
import io.realm.RealmConfiguration
import java.util.*
import javax.inject.Named
import javax.inject.Singleton

/**
 * @author Martin Macheiner
 * Date: 14.06.2017.
 */
@Module
class AppModule(private val app: Application) {

    @Provides
    @Singleton
    fun provideCurrencyManager(priceProxy: PriceProxy,
                               storageManager: StorageManager,
                               currencyConversionProvider: CurrencyConversionProvider,
                               preferences: SharedPreferences): CurrencyManager {
        return DefaultCurrencyManager(priceProxy, storageManager,
                currencyConversionProvider, preferences)
    }

    @Provides
    @Singleton
    fun providePriceProxy(@Named("CoinbasePriceProvider") coinbaseProvider: PriceProvider,
                          @Named("CoinMarketCapPriceProvider") coinMarketCapProvider: PriceProvider): PriceProxy {
        return DefaultPriceProxy(Arrays.asList(coinbaseProvider, coinMarketCapProvider))
    }

    @Provides
    @Singleton
    fun provideRealm(): Realm {
        val config = RealmConfiguration.Builder()
                .schemaVersion(CryptoWatcherRealmMigration.v1_0_scheme)
                .migration(CryptoWatcherRealmMigration())
                //.encryptionKey(key)
                .build()
        return Realm.getInstance(config)
    }

    @Provides
    @Singleton
    fun provideStorageManager(realm: Realm): StorageManager {
        return RealmStorageManager(realm)
    }

    @Provides
    @Singleton
    fun provideSharedPreferences(): SharedPreferences {
        return PreferenceManager.getDefaultSharedPreferences(app.applicationContext)
    }

}
